### ISSUE TITLE
Need findROCM with modern CMake support

### ISSUE THREAD

**Issue by ye-luo** [ROCm/ROCm#1259]
Current CMake needs improvements. Please match [FindCUDAToolkit](https://cmake.org/cmake/help/git-stage/module/FindCUDAToolkit.html) and define targets for all the components.
FindHIP and FindROCR are too basic and pre-modern cmake.

**Comment by rkothako** [comment]
Hi @ye-luo 
Thanks for the ticket.
We are checking with corresponding team internally and will get back to you soon.

**Comment by rkothako** [comment]
Hi @ye-luo 
We have plans to upgrade to modern cmake for ROCm but discussions are in progress.
We will inform once we upgrade officially.

**Comment by rkothako** [comment]
I request you close this ticket meanwhile.

**Comment by ye-luo** [comment]
I will close this ticket when the first bits of progress shows up in the release.

**Comment by ye-luo** [comment]
My first ride is pretty rough. Here is the list of issues to track
https://github.com/ROCm-Developer-Tools/HIP/issues/2433
https://github.com/ROCm-Developer-Tools/HIP/pull/2434
https://github.com/ROCm-Developer-Tools/hipamd/pull/7
https://github.com/RadeonOpenCompute/ROCm/issues/1636
https://gitlab.kitware.com/cmake/cmake/-/issues/23006

My goal is to get the following working
```
  enable_language(HIP)
  list(PREPEND CMAKE_PREFIX_PATH ${CMAKE_HIP_COMPILER_ROCM_ROOT})
  find_package(hip CONFIG REQUIRED)
  find_package(hipblas CONFIG REQUIRED)
```
but it seems needing multiple fixes. I hope to someone from AMD who has a extensive picture of CMake in ROCm to look at those.

**Comment by ROCmSupport** [comment]
Hi @ye-luo 
As discussed in the previous comments, hope you are going to close this ticket. Thank you.

**Comment by Maxzor** [comment]
Please do worry less about closing this ticket, than about fixing the current cmake/hipconfig situation.
I am doing an effort for native Debian packaging, and the general assumption that the ROCm toolkit is all under one directory - binaries, libraries, be it in the default /opt/rocm or somewhere else... feels pretty wrong.

The hipconfig and hipvars scripts are especially weak.

I am trying to get the hip cmake test suite (on top of the external [test-suite](https://github.com/rocm-developer-tools/hip-testsuite) and [example](https://github.com/rocm-developer-tools/hip-examples) repos) working with a decoupled install. ROCm libraries in /usr/{lib, share}/rocm and binaries in /usr/bin/rocm : I am currently fighting the whole build system of ROCm and this is exhausting. **This person's work really looks like going in the good direction**, into making more robust and less convoluted design decisions.

**Comment by ye-luo** [comment]
I just noticed another trouble https://github.com/RadeonOpenCompute/ROCm/issues/1717

**Comment by torehl** [comment]
@Maxzor 

> I am doing an effort, alongside the dedicated packaging team, for native Debian packaging, and the general assumption that the ROCm toolkit is all under one directory - binaries, libraries, be it in the default /opt/rocm or somewhere else... feels pretty wrong.

I completely agree. - Been trying to make ROCm work in a module environment, and think this needs fixing in order for it to work properly. Also opened https://gitlab.freedesktop.org/drm/amd/-/issues/1961 in hope of getting them to allow for "--prefix" and "--modulefiles" to generate  modules files for ROCm install.

I tried tar'ing to keep soft links and moved it to /cm/shared/apps/amd/rocm/5.1.0 (Bright (soon Nvidia) CM), and configured a modules file, the enduser could compile on login node w/o AMD GPUs. I could also build targeted clang and gcc. But during runtime, it seems to have issues picking up soflinked dso's of underlying directories, and I think this is related to what you describe. Looks like I need to copy all dso's into the main lib directory in the same way Nvidia cuda has it.

**Comment by keryell** [comment]
> Hi @ye-luo As discussed in the previous comments, hope you are going to close this ticket. Thank you.

@ROCmSupport Is the problem solved?

**Comment by cgmb** [comment]
@torehl, if you haven't checked it out before, maybe give [Spack](https://spack.readthedocs.io) a try. AMD contributes build recipes to Spack for much of the ROCm stack, and [it integrates with lmod](https://spack.readthedocs.io/en/latest/module_file_support.html).

**Comment by ronlieb** [comment]
@ROCmSupport   please address this issue

**Comment by nartmada** [comment]
Sorry for the delay @ronlieb and @ye-luo.  I had a chat with the internal team and they said the support has been implemented in cmake 3.21.X and available to the public.  

https://cmake.org/cmake/help/latest/search.html?q=hip

Please check and let me know if this is still an issue.  Thanks for your patience.

**Comment by ye-luo** [comment]
Working fine now.
